docs/cookbook/extending-the-ui.md

```diff
@@ -0,0 +1,990 @@
+    flex-direction: row;
+
+    .apos-schema {
+      flex: 2;
+      margin-right: 10px;
+      display: flex;
+      flex-direction: row;
+
+      [data-apos-field] {
+        margin-left: 10px;
+
+        &:first-child {
+          margin-left: 0;
+        }
+      }
+    }
+
+    .letter-grade {
+      flex: 1;
+      display: flex;
+      align-items: center;
+      justify-content: center;
+      font-size: 2rem;
+      font-weight: bold;
+      margin-bottom: 30px;
+    }
+  }
+}
+</style>
+```
+
+<template v-slot:caption>
+/modules/grade/ui/apos/components/InputGradeField.vue
+</template>
+
+</AposCodeBlock>
+
+**The schema field markup**
+
+```javascript
+<AposInputWrapper :field="field" :error="null" :uid="uid" :modifiers="modifiers">
+...
+</AposInputWrapper>
+```
+
+For our custom schema field, we are utilizing two helper components from the `@apostrophecms/schema` module. The first is the `AposInputWrapper`. This component is used for error handling and input appearance, as well as behind-the-scenes housekeeping. It wraps the entire code block where the Apostrophe schema fields are being used. Four props are typically passed to this component.
+
+The only one that we might want to change the value of is `:error`. It can take values of either `"null"` or `"effectiveError"`. If you are making a component that has multiple schema fields and set `:error` to "null", errors will only be shown for the individual fields, not the top-level component. Setting it to "effectiveError" will show errors in the individual fields, as well as the entire custom schema field. It is a judgment call to determine which will result in a better user experience.
+
+:::note
+As a guide for selecting error type, the Apostrophe `array` and `object` fields use `:error="null"`. This provides clear guidance on which field is in error, without a second error message being displayed for the entire component. You may want to change this if your custom field has a requirement such as `min: 1`.
+:::
+
+```javascript
+<template #body>
+  <div class="custom-input-object">
+    <div class="custom-input-wrapper">
+      <AposSchema :schema="gradeSchema" :trigger-validation="triggerValidation" :utility-rail="false" :generation="generation" v-model="gradeSchemaInput">
+      </AposSchema>
+      <span class="letter-grade">{{ letterGrade() }}</span>
+    </div>
+  </div>
+</template>
+```
+
+The `AposSchema` component does the majority of work constructing and displaying our schema fields, as well as passing data.
+
+In this case, we are wrapping it in some additional markup to allow for customized styling. In making your components, this may not be necessary. To return the letter grade we are adding an additional span after the `AposSchema` with content set by the `letterGrade()` method.
+
+The `AposSchema` component typically takes the `:trigger-validation`, `:utility-rail` and `:generation` props. These are used by Apostrophe for field validation, display, and updating.
```

Comment:
`utility-rail` should default to false. It's such low-hanging fruit, for sure. But, that's another PR.

Reply:
... I just sent you a core PR for that, so that starting with our next release you won't have to document the silly requirement to pass `utility-rail` explicitly.

Reply:
Updated